### PR TITLE
[Pools] Ensure members can always exit the pool gracefully

### PR DESCRIPTION
**Pull request by Ank4n**
Resolves https://github.com/paritytech-secops/srlabs_findings/issues/412

## Changes
- Clear any dangling delegation when member is removed.
- Agents need to be killed explicitly when pools are destroyed.
- Member withdraw amount is max of their locked funds and the value of their points.

### CODE REVIEW

File: substrate/frame/nomination-pools/src/lib.rs [hunk 1]
```diff
@@ -1831,7 +1831,9 @@ pub mod pallet {
 		/// A member has been removed from a pool.
 		///
 		/// The removal can be voluntary (withdrawn all unbonded funds) or involuntary (kicked).
-		MemberRemoved { pool_id: PoolId, member: T::AccountId },
+		/// Any funds that was still delegated (dangling delegation) is cleared and is represented
```
[the remark this answers is not in the document]
Comment (Ank4n) [reply substrate/frame/nomination-pools/src/lib.rs:1834]: Technically it is the hold that is getting cleared (not withdrawn). But may be its still better to call it a withdraw?

File: substrate/frame/nomination-pools/src/lib.rs [hunk 2]
```diff
@@ -1831,7 +1831,9 @@ pub mod pallet {
...
+		/// by `balance`.
+		MemberRemoved { pool_id: PoolId, member: T::AccountId, balance: BalanceOf<T> },
```
[the remark this answers is not in the document]
Comment (Ank4n) [reply substrate/frame/nomination-pools/src/lib.rs:1836]: The balance is not really reaped right? It is just released.

File: substrate/frame/delegated-staking/src/lib.rs [hunk 3]
```diff
@@ -599,44 +620,19 @@ impl<T: Config> Pallet<T> {
...
+		}
 
 		// if we still do not have enough funds to release, abort.
 		ensure!(agent_ledger.ledger.unclaimed_withdrawals >= amount, Error::<T>::NotEnoughFunds);
+		agent_ledger.remove_unclaimed_withdraw(amount)?.update();
 
-		// Claim withdraw from agent. Kill agent if no delegation left.
-		// TODO: Ideally if there is a register, there should be an unregister that should
-		// clean up the agent. Can be improved in future.
-		if agent_ledger.remove_unclaimed_withdraw(amount)?.update_or_kill()? {
```
[the remark this answers is not in the document]
Comment (Ank4n) [reply substrate/frame/delegated-staking/src/lib.rs:621]: It is cleaner to remove agent explicitly when the pool is dissolved. We don't want to end up in a state where pool exists but agent dropped.

File: substrate/frame/nomination-pools/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gpestana) [reply substrate/frame/nomination-pools/src/lib.rs]: Yes, I think released is the best way to put it.

File: substrate/frame/nomination-pools/src/adapter.rs [hunk 4]
```diff
@@ -300,6 +308,17 @@ impl<T: Config, Staking: StakingInterface<Balance = BalanceOf<T>, AccountId = T:
...
 	}
 
+	fn dissolve(pool_account: Pool<Self::AccountId>) -> DispatchResult {
+		defensive_assert!(
+			T::Currency::total_balance(&pool_account.clone().get()).is_zero(),
+			"dissolving pool should not have any balance"
+		);
+
+		// Defensively force set balance to zero.
+		T::Currency::set_balance(&pool_account.get(), Zero::zero());
```
Comment (kianenigma) [inline substrate/frame/nomination-pools/src/adapter.rs:318]: Does this update the TI?

**Review by kianenigma — APPROVED** [review]

File: substrate/frame/nomination-pools/src/adapter.rs (hunk 4 above)
Comment (Ank4n) [reply substrate/frame/nomination-pools/src/adapter.rs:318]: It does. Also this is for the older `TransferStake` code, just refactored and the logic stays same.

File: substrate/frame/delegated-staking/src/lib.rs [hunk 5]
```diff
@@ -304,6 +304,27 @@ pub mod pallet {
...
+
+			ensure!(
+				ledger.total_delegated == Zero::zero() &&
+					ledger.pending_slash == Zero::zero() &&
+					ledger.unclaimed_withdrawals == Zero::zero(),
+				Error::<T>::NotAllowed
+			);
+
+			// remove provider reference
+			let _ = frame_system::Pallet::<T>::dec_providers(&who).defensive();
```
Comment (gpestana) [inline substrate/frame/delegated-staking/src/lib.rs:323]: I'm wondering if this is the best way to deal with a potential error returned from `dec_providers`. Maybe we should return earlier and not remove the agent entry from the map unless the providers were decreased?

OTOH, if there's an error, the agent will linger and the pool may not be destroyed, so probably the best way is to keep this defensive and then fix the inconsistency with a simple migration or something along those lines. wdyt?

File: substrate/frame/nomination-pools/src/lib.rs [hunk 6]
```diff
@@ -1831,7 +1831,9 @@ pub mod pallet {
 		/// A member has been removed from a pool.
 		///
 		/// The removal can be voluntary (withdrawn all unbonded funds) or involuntary (kicked).
-		MemberRemoved { pool_id: PoolId, member: T::AccountId },
+		/// Any funds that was still delegated (dangling delegation) is released and is represented
+		/// by `released_balance`.
```
Comment (gpestana) [inline substrate/frame/nomination-pools/src/lib.rs:1835]:
```suggestion
		/// Any funds that are still delegated (i.e. dangling delegation) are released and are represented
		/// by `released_balance`.
```

**Review by gpestana — APPROVED** [review]

**Comment by Ank4n** [comment]
bot fmt

**Pull request merged**